compiler/rustc_smir/src/rustc_smir/mod.rs

```diff
@@ -79,6 +80,57 @@ impl<'tcx> Tables<'tcx> {
         };
         !must_override && self.tcx.is_mir_available(def_id)
     }
+
+    fn to_fn_def(&mut self, tcx: TyCtxt<'tcx>, def_id: DefId) -> Option<FnDef> {
+        if matches!(tcx.def_kind(def_id), DefKind::Fn | DefKind::AssocFn) {
+            Some(self.fn_def(def_id))
+        } else {
+            None
+        }
+    }
+
+    fn to_static(&mut self, tcx: TyCtxt<'tcx>, def_id: DefId) -> Option<StaticDef> {
+        matches!(tcx.def_kind(def_id), DefKind::Static { .. }).then(|| self.static_def(def_id))
+    }
+
+    fn filter_map_items<T, F>(&mut self, tcx: TyCtxt<'tcx>, krate: CrateNum, func: F) -> Vec<T>
+    where
+        F: Fn(&mut Tables<'tcx>, TyCtxt<'tcx>, DefId) -> Option<T>,
+    {
+        let crate_def = rustc_span::def_id::DefId { index: CRATE_DEF_INDEX, krate };
+        let mut queue = vec![crate_def];
+        let mut visited = HashSet::default();
+        let mut result = vec![];
+        while let Some(next) = queue.pop() {
+            for def_id in tcx.module_children(next).iter().filter_map(|item| item.res.opt_def_id())
+            {
+                if !visited.contains(&def_id) {
+                    visited.insert(def_id);
+                    result.extend(func(self, tcx, def_id));
+                    match tcx.def_kind(def_id) {
+                        DefKind::Mod | DefKind::ForeignMod => queue.push(def_id),
+                        DefKind::Struct | DefKind::Enum | DefKind::Union => {
+                            for associated_item in tcx
+                                .inherent_impls(def_id)
+                                .iter()
+                                .flat_map(|impl_id| tcx.associated_item_def_ids(impl_id))
+                            {
+                                result.extend(func(self, tcx, *associated_item));
+                            }
+                        }
+                        DefKind::Trait => {
+                            for associated_item in tcx.associated_item_def_ids(def_id) {
+                                result.extend(func(self, tcx, *associated_item));
+                            }
+                        }
+                        _ => {}
```

Comment:
Why aren't you just handling `DefKind::Impl` here?

Reply:
Then you won't need to iterate through the inherent impls above, and you get all the trait implementations too.

Reply:
You should also just be able to push traits and impls onto the queue. They have module children.

Reply:
I thought I tried that, but I might've used the wrong implementation. I might've tried the associated functions not module children. Let me try it. Thanks!

Reply:
I tried the following:
```rust
        while let Some(def_id) = queue.pop() {
            if visited.insert(def_id) {
                result.extend(func(self, tcx, def_id));
                match tcx.def_kind(def_id) {
                    DefKind::Mod | DefKind::ForeignMod | DefKind::Impl { .. } | DefKind::Trait => {
                        queue.extend(
                            tcx.module_children(def_id)
                                .iter()
                                .filter_map(|item| item.res.opt_def_id()),
                        )
                    }
                    _ => {}
                }
            }
        }
```

But in that case, I can no longer find method implementations, such as `std::option::Option::<T>::is_some`. Trait implementations are still not showing up. What am I doing wrong?

Reply:
I will take a look when I am at my computer.... can you push some code to test this?